drivers/dma/dma_nios2_msgdma.c

```diff
@@ -0,0 +1,240 @@
+#include <dma.h>
+#include <altera_common.h>
+#include "altera_msgdma_csr_regs.h"
+#include "altera_msgdma_descriptor_regs.h"
+#include "altera_msgdma.h"
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_DMA_LEVEL
+#include <logging/sys_log.h>
+
+/* Device configuration parameters */
+struct nios2_msgdma_dev_cfg {
+	alt_msgdma_dev *msgdma_dev;
+	alt_msgdma_standard_descriptor desc;
+	u32_t direction;
+	struct k_sem sem_lock;
+	void (*dma_callback)(struct device *dev, u32_t id,
+			     int error_code);
+};
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_CFG(dev) \
+	((struct nios2_msgdma_dev_cfg *)(dev)->config->config_info)
+
+static void nios2_msgdma_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct nios2_msgdma_dev_cfg *cfg = DEV_CFG(dev);
+
+	/* Call Altera HAL driver ISR */
+	alt_handle_irq(cfg->msgdma_dev, MSGDMA_0_CSR_IRQ);
+}
+
+static void nios2_msgdma_callback(void *context)
+{
+	struct nios2_msgdma_dev_cfg *dev_cfg =
+				DEV_CFG((struct device *)context);
+	int err_code;
+	u32_t status;
+
+	status = IORD_ALTERA_MSGDMA_CSR_STATUS(dev_cfg->msgdma_dev->csr_base);
+
+	if (status & ALTERA_MSGDMA_CSR_STOPPED_ON_ERROR_MASK) {
+		err_code = -EIO;
+	} else if (status & ALTERA_MSGDMA_CSR_BUSY_MASK) {
+		err_code = -EBUSY;
+	} else {
+		err_code = 0;
+	}
+
+	SYS_LOG_DBG("msgdma csr status Reg: 0x%x", status);
+
+	dev_cfg->dma_callback((struct device *)context, 0, err_code);
+}
+
+static int nios2_msgdma_config(struct device *dev, u32_t channel,
+			       struct dma_config *cfg)
+{
+	struct nios2_msgdma_dev_cfg *dev_cfg = DEV_CFG(dev);
+	struct dma_block_config *dma_block;
+	int status;
+	u32_t control;
+
+	/* Nios-II MSGDMA supports only one channel per DMA core */
+	if (channel != 0) {
+		SYS_LOG_ERR("invalid channel number");
+		return -EINVAL;
+	}
+
+#if MSGDMA_0_CSR_PREFETCHER_ENABLE
+	if (cfg->block_count > 1) {
+		SYS_LOG_ERR("driver yet add support multiple descriptors");
+		return -EINVAL;
+	}
+#else
+	if (cfg->block_count != 1) {
+		SYS_LOG_ERR("invalid block count!!");
+		return -EINVAL;
+	}
+#endif
+
+	if (cfg->head_block == NULL) {
+		SYS_LOG_ERR("head_block ptr NULL!!");
+		return -EINVAL;
+	}
+
+	if (cfg->head_block->block_size > MSGDMA_0_DESCRIPTOR_SLAVE_MAX_BYTE) {
+		SYS_LOG_ERR("DMA error: Data size too big: %d",
+			    cfg->head_block->block_size);
+		return -EINVAL;
+	}
+
+	k_sem_take(&dev_cfg->sem_lock, K_FOREVER);
+	dev_cfg->dma_callback = cfg->dma_callback;
+	dev_cfg->direction = cfg->channel_direction;
+	dma_block = cfg->head_block;
+	control =  ALTERA_MSGDMA_DESCRIPTOR_CONTROL_TRANSFER_COMPLETE_IRQ_MASK |
+		  ALTERA_MSGDMA_DESCRIPTOR_CONTROL_EARLY_TERMINATION_IRQ_MASK;
+
+	/*
+	 * Suppress -Wint-conversion compiler warnings which
+	 * are caused due to the type mismatch b/w the dma_block
+	 * source/destination address (u32_t) and HAL driver's
+	 * descriptor construction API's arguments (alt_32 *)
+	 */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wint-conversion"
```

Comment:
Why do it this way, instead of just casting?

Reply:
simple (u32_t *) didn't work, now i have changed it to (alt_u32 *) and it works fine.
I will remove the prama macros. Thanks.